### PR TITLE
Fix changelog generation

### PR DESCRIPTION
**Pull request by JounQin**
> Which issue, if any, is this issue related to?

Related #6282

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

### CODE REVIEW

File: .changeset/changelog-stylelint.js
Comment (JounQin) [inline .changeset/changelog-stylelint.js:24]: I'm still not quite sure how should I implement this order algorithm, @jeddy3 @ybiquitous can you offer some help here?

File: patches/@changesets+apply-release-plan+6.0.4.patch
Comment (JounQin) [inline patches/@changesets+apply-release-plan+6.0.4.patch:19]: I'd like PR back to changesets if we use this function successfully.

File: package.json
Comment (JounQin) [inline package.json:198]: This fixed https://github.com/stylelint/stylelint/pull/6282#discussion_r949054860

File: .changeset/changelog-stylelint.js
Comment (ybiquitous) [reply .changeset/changelog-stylelint.js:24]: If it's hard to implement the order, it seems that we can manually fix the generated changelog entries. What do you think?

File: package.json
Comment (ybiquitous) [reply package.json:198]: Could you explain why this `overrides.prettier` is necessary?

File: .changeset/changelog-stylelint.js
Comment (JounQin) [reply .changeset/changelog-stylelint.js:24]: It's not hard, I just don't know what should I write.  😅

File: package.json
Comment (JounQin) [reply package.json:198]: > Could you explain why this `overrides.prettier` is necessary?

There are different prettier versions installed. `^1. 19` vs `^2. 0` which  could have different defaults O suppose.

File: patches/@changesets+apply-release-plan+6.0.4.patch
Comment (JounQin) [reply patches/@changesets+apply-release-plan+6.0.4.patch:1]: See https://github.com/stylelint/stylelint/pull/6299#discussion_r952719715

File: .changeset/changelog-stylelint.js
Comment (JounQin) [reply .changeset/changelog-stylelint.js:24]: @ybiquitous I don't understand

> Order the item within the group by the widest-reaching first to the smallest, and then alphabetically by rule name.

What does this mean? Sorry, I'm not a native English speaker.

Comment (JounQin) [reply .changeset/changelog-stylelint.js]: Can you help to review whether my latest changes are expected?

File: package.json
Comment (ybiquitous) [reply package.json:198]: Is this the same problem as https://github.com/changesets/changesets/issues/616?

File: patches/@changesets+apply-release-plan+6.0.4.patch
Comment (ybiquitous) [reply patches/@changesets+apply-release-plan+6.0.4.patch:1]: I did a quick search for issues or pull requests to request such a feature in the [changesets/changesets](https://github.com/changesets/changesets) repo, but I could not... 😓 Do you have any information?

Honestly, I'm worried that we may need to test or modify this patch with each update of the `@changeset/*` packages. At least, I wish there would be a feature request issue and a pull request using someone's fork version, instead of applying and managing the patch in our repository.

File: package.json
Comment (JounQin) [reply package.json:198]: Seems true.

File: patches/@changesets+apply-release-plan+6.0.4.patch
Comment (JounQin) [reply patches/@changesets+apply-release-plan+6.0.4.patch:1]: When we agree with these changes and merge, I'll pr back, I want to guarded it works as expected for us first.

Comment (JounQin) [reply patches/@changesets+apply-release-plan+6.0.4.patch:1]: @ybiquitous I just raised https://github.com/changesets/changesets/issues/926

File: package.json
Comment (ybiquitous) [reply package.json:198]: What about ignoring `CHANGELOG.md` like [this workaround](https://github.com/PrairieLearn/PrairieLearn/commit/a9b1af0f21d7e55339533d7bf364545f40011b51)? Because it's unnecessary to change dependencies and `package.json` and possible to leave a TODO comment, in addition likely, we won't modify `CHANGELOG.md` manually with the new workflow.

Comment (JounQin) [reply package.json:198]: Ignoring will result inconsistent list item maybe, but personally I want a single prettier installed to reduce related issues. https://github.com/PrairieLearn/PrairieLearn/commit/a9b1af0f21d7e55339533d7bf364545f40011b51 this workaround Is not good at all IMO.

And also, Is there any downside by using `overrides`?

Comment (ybiquitous) [reply package.json:198]: It's challenging to find the right place. 🤔 

I'm not sure about any downsides with `overrides`, but I have a little worry about dependency management; whether `overrides` may cause other problems.

Comment (JounQin) [reply package.json:198]: I don't see any downside by overriding `prettier`'s version, it should always be consistent, otherwise different prettier may depend on different prettier version which results in unexpected differences like this due to inconsistent defaults.

Comment (ybiquitous) [reply package.json:198]: Sounds good. 👍🏼 
When `chagesets` fixes the problem, let's revert this `overrides` change.

**Comment by JounQin** [comment]
@ybiquitous @jeddy3 Are we good to go forward.

File: .changeset/changelog-stylelint.js
Comment (ybiquitous) [reply .changeset/changelog-stylelint.js]: Could you let us know how to test this change locally?

Comment (JounQin) [reply .changeset/changelog-stylelint.js]: @ybiquitous 

1. generate some changesets files with `npx changeset`
2. run `npm run version` and check CHANGELOG file and `package.json`

Comment (ybiquitous) [reply .changeset/changelog-stylelint.js]: I've tested it locally just to run the `npm run version` command, and then the diff is below:

```diff
diff --git a/.changeset/nice-lemons-reflect.md b/.changeset/nice-lemons-reflect.md
deleted file mode 100644
index 6f75e4c2c..000000000
--- a/.changeset/nice-lemons-reflect.md
+++ /dev/null
@@ -1,5 +0,0 @@
----
-"stylelint": minor
----
-
-Added: `declaration-block-no-duplicate-properties` autofix
diff --git a/CHANGELOG.md b/CHANGELOG.md
index d957c5985..e3cd7e630 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,11 @@
 # Changelog
 
+## 14.12.0
+
+### Minor Changes
+
+- Added: `declaration-block-no-duplicate-properties` autofix ([#6296](https://github.com/stylelint/stylelint/pull/6296)) ([@fpetrakov](https://github.com/fpetrakov)).
+
 ## 14.11.0
 
 - Added: `ignoreAfterCombinators: []` to `selector-max-universal` ([#6275](https://github.com/stylelint/stylelint/pull/6275)).
```

The `### Minor Changes` section seems extra, but we can manually remove it on a release pull request.

File: .changeset/changelog-stylelint.js
Comment (ybiquitous) [inline .changeset/changelog-stylelint.js:10]: [suggest] How about refactoring the variable like this? If doing so, `@type {Array<string | undefined>}` also may be no longer needed.

```js
const CHANGESET_SECTIONS = ['Deprecated', 'Removed', 'Changed', 'Added', 'Fixed'];
```

File: .changeset/changelog-stylelint.js
Comment (ybiquitous) [reply .changeset/changelog-stylelint.js]: As the [maintainer guide describes](https://stylelint.io/maintainer-guide/pull-requests/#merging) as below,

> Prefix the item with either: "Removed", "Changed", "Deprecated", "Added", or "Fixed".

Is it possible to sort change entries by order of `"Removed", "Changed", "Deprecated", "Added", or "Fixed"` (importance order)?

- `"Removed"`, `"Changed"` - major (breaking)
- `"Deprecated"`, `"Added"` - minor
- `"Fixed"` - patch

Comment (JounQin) [reply .changeset/changelog-stylelint.js]: I remembered you said, but I didn't find the source: we can endure these changes. And the reorder algorithm only works in sections.

Besides, such section titles are much easier for reading IMO.

Even if we want to rollback, we can just remove these section title.

File: .changeset/changelog-stylelint.js
Comment (JounQin) [reply .changeset/changelog-stylelint.js:10]: Working on it.

File: .changeset/changelog-stylelint.js
Comment (JounQin) [reply .changeset/changelog-stylelint.js]: Personally, I don't think changesets will officially support this, but fork to support our purpose is very simple just like our patch right now.

Comment (JounQin) [reply .changeset/changelog-stylelint.js]: I also don't quite sure why we're focusing these unimportant things.🤣

Comment (JounQin) [reply .changeset/changelog-stylelint.js]: I have an idea how to implement it just now, I'll try ASAP.

**Comment by JounQin** [comment]
friendly ping @ybiquitous @jeddy3

**Review by ybiquitous — CHANGES_REQUESTED** [review]
When I run `npm run version`, the extra sections are not inserted. Thank you.

Instead, the `- patch` line is inserted. Here is an example of `CHANGELOG.md` diff:

```diff
@@ -1,5 +1,10 @@
 # Changelog
 
+## 14.12.0
+
+- patch
+- Added: `declaration-block-no-duplicate-properties` autofix ([#6296](https://github.com/stylelint/stylelint/pull/6296)) ([@fpetrakov](https://github.com/fpetrakov)). - minor
+
 ## 14.11.0
 
 - Added: `ignoreAfterCombinators: []` to `selector-max-universal` ([#6275](https://github.com/stylelint/stylelint/pull/6275)).
```

**Comment by JounQin** [comment]
> Instead, the - patch line is inserted. Here is an example of CHANGELOG.md diff:

Interesting, let me check.

File: .changeset/changelog-stylelint.js
Comment (ybiquitous) [inline .changeset/changelog-stylelint.js:33]: [suggest] I noticed an extra suffix `- minor` at the end of the line below when running `npm run version`:

```diff
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,9 @@
 # Changelog
 
+## 14.12.0
+
+- Added: `declaration-block-no-duplicate-properties` autofix ([#6296](https://github.com/stylelint/stylelint/pull/6296)) ([@fpetrakov](https://github.com/fpetrakov)). - minor
+
 ## 14.11.0
 
 - Added: `ignoreAfterCombinators: []` to `selector-max-universal` ([#6275](https://github.com/stylelint/stylelint/pull/6275)).
```

Can the following change fix it?

```suggestion
						acc.push(`${line}${type === 'major' ? ' (BREAKING)' : ''}`);
```

Comment (JounQin) [reply .changeset/changelog-stylelint.js:33]: So I misunderstood your comment before. 🤣

File: .changeset/changelog-stylelint.js
Comment (ybiquitous) [inline .changeset/changelog-stylelint.js:10]: [suggest] Please use the importance order described in the [doc](https://stylelint.io/maintainer-guide/pull-requests#merging):

> Prefix the item with either: "Removed", "Changed", "Deprecated", "Added", or "Fixed".

```suggestion
const CHANGESET_SECTIONS = ['Removed', 'Changed', 'Deprecated', 'Added', 'Fixed'];
```

File: .changeset/changelog-stylelint.js
Comment (ybiquitous) [inline .changeset/changelog-stylelint.js:12]: [suggest] `- ` at the begin of a line seems necessary:

```suggestion
const changesetSectionReg = new RegExp(`^- (${CHANGESET_SECTIONS.join('|')}): `);
```

File: .changeset/changelog-stylelint.js
Comment (ybiquitous) [inline .changeset/changelog-stylelint.js:40]: [suggest] `1` seems correct, instead of `0`:

```suggestion
				const aSection = changesetSectionReg.exec(a)?.[1];
				const bSection = changesetSectionReg.exec(b)?.[1];
```

File: .changeset/changelog-stylelint.js
Comment (ybiquitous) [inline .changeset/changelog-stylelint.js:43]: [sugget] The order seems incorrect:

```suggestion
					? a.localeCompare(b)
					: CHANGESET_SECTIONS.indexOf(aSection) - CHANGESET_SECTIONS.indexOf(bSection);
```

**Review by ybiquitous — APPROVED** [review]
Thanks. LGTM 👍🏼 

I locally confirm it works:

```diff
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,16 @@
 # Changelog
 
+## 15.0.0
+
+- Removed: `foo` rule. (BREAKING)
+- Changed: `foo` rule. (BREAKING)
+- Deprecated: `foo` rule.
+- Added: `bar` rule.
+- Added: `declaration-block-no-duplicate-properties` autofix ([#6296](https://github.com/stylelint/stylelint/pull/6296)) ([@fpetrakov](https://github.com/fpetrakov)).
+- Added: `foo` rule.
+- Fixed: `bar` rule.
+- Fixed: `foo` rule.
+
 ## 14.11.0
 
 - Added: `ignoreAfterCombinators: []` to `selector-max-universal` ([#6275](https://github.com/stylelint/stylelint/pull/6275)).
```

**Comment by JounQin** [comment]
@ybiquitous Let's merge? cc @jeddy3

**Comment by ybiquitous** [comment]
I'm curious about the other members' thoughts on the patch and the tricky script, so let's wait for feedback a little longer.

By the way, I notice https://github.com/changesets/changesets/issues/616 was closed today. Is `overrides.prettier` still necessary? (ref: https://github.com/stylelint/stylelint/pull/6299#discussion_r952737946)

**Comment by JounQin** [comment]
> I'm curious about the other members' thoughts on the patch and the tricky script, so let's wait for feedback a little longer.

It seems there are only us two invoked in this PR.

> By the way, I notice https://github.com/changesets/changesets/issues/616 was closed today. Is overrides.prettier still necessary? (ref: https://github.com/stylelint/stylelint/pull/6299#discussion_r952737946)

Nice catch, I'll check ASAP.

**Comment by JounQin** [comment]
> Could you fix the diff also, please?

Removed, although why it is added previously. 😂

**Comment by JounQin** [comment]
I'm going to merge and wait for feedback from the other maintainers.

**Pull request merged**
